File: src/main/java/de/weimarnetz/registrator/services/NodeNumberService.java
package de.weimarnetz.registrator.services;

import java.util.List;
import java.util.stream.Collectors;

import org.apache.commons.lang3.tuple.Pair;
import org.springframework.stereotype.Component;

import de.weimarnetz.registrator.configuration.NetworksConfiguration;
import de.weimarnetz.registrator.exceptions.NetworkNotFoundException;
import de.weimarnetz.registrator.exceptions.NoMoreNodesException;
import de.weimarnetz.registrator.model.Node;
import de.weimarnetz.registrator.repository.RegistratorRepository;

import lombok.AllArgsConstructor;

@Component
@AllArgsConstructor
public class NodeNumberService {

    private final RegistratorRepository registratorRepository;
    private final NetworksConfiguration networksConfiguration;

    public int getNextAvailableNodeNumber(String network) throws NoMoreNodesException, NetworkNotFoundException {
        Pair<Integer, Integer> nodeNumberBoundaries = getNodeNumberBoundaries(network);
        List<Node> nodeList = registratorRepository.findAllByNetwork(network);
        List<Integer> nodeNumbersList = nodeList.stream()
                .map(Node::getNumber)
                .sorted()
                .collect(Collectors.toList());
        if (nodeNumbersList.size() > nodeNumberBoundaries.getRight() - nodeNumberBoundaries.getLeft()) {
            throw new NoMoreNodesException();
        }
        return findFirstMissing(nodeNumbersList, 0, nodeNumbersList.size() - 1, network);

    }

    public boolean isNodeNumberValid(int nodeNumber, final String network) throws NetworkNotFoundException {
        return nodeNumber >= getNodeNumberBoundaries(network).getLeft() && nodeNumber <= getNodeNumberBoundaries(network).getRight();
    }

    private int findFirstMissing(List<Integer> list, int start, int end, String network) throws NetworkNotFoundException {
        int minNodeNumber = getNodeNumberBoundaries(network).getLeft();
        if (list.isEmpty()) {
            return minNodeNumber;
        }

        if (start > end) {
            return end + 1 + minNodeNumber;
        }

        if (start + minNodeNumber != list.get(start)) {
            return start + minNodeNumber;
        }

        int mid = (start + end) / 2;

        // Left half has all elements from 0 to mid
        if (list.get(mid) == mid + minNodeNumber) {
            return findFirstMissing(list, mid + 1, end, network);
        }

        return findFirstMissing(list, start, mid, network);
    }

    private Pair<Integer, Integer> getNodeNumberBoundaries(String network) throws NetworkNotFoundException {
        if (networksConfiguration.getMap().containsKey(network)) {
            return Pair.of(networksConfiguration.getMap().get(network).getMinNodeNumber(),
                    networksConfiguration.getMap().get(network).getMaxNodeNumber());
        } else {
            throw new NetworkNotFoundException("Network " + network + " doesn't exist");
        }
    }

}

File: src/main/java/de/weimarnetz/registrator/services/NodeNumberService.kt
package de.weimarnetz.registrator.services

import de.weimarnetz.registrator.configuration.NetworksConfiguration
import de.weimarnetz.registrator.exceptions.NetworkNotFoundException
import de.weimarnetz.registrator.exceptions.NoMoreNodesException
import de.weimarnetz.registrator.repository.RegistratorRepository
import org.springframework.stereotype.Component

@Component
class NodeNumberService(
    private val registratorRepository: RegistratorRepository,
    private val networksConfiguration: NetworksConfiguration
) {
    @Throws(NoMoreNodesException::class, NetworkNotFoundException::class)
    fun getNextAvailableNodeNumber(network: String): Int {
        val nodeNumberBoundaries = getNodeNumberBoundaries(network)
        val nodeList = registratorRepository.findAllByNetwork(network)
        val nodeNumbersList = nodeList
            .map { it.number }
            .sorted()
            .toList()
        if (nodeNumbersList.size > nodeNumberBoundaries.second - nodeNumberBoundaries.first) {
            throw NoMoreNodesException()
        }
        return findFirstMissing(nodeNumbersList, 0, nodeNumbersList.size - 1, network)
    }

    @Throws(NetworkNotFoundException::class)
    fun isNodeNumberValid(nodeNumber: Int, network: String?): Boolean {
        return nodeNumber >= getNodeNumberBoundaries(network).first && nodeNumber <= getNodeNumberBoundaries(network).second
    }

    private fun findFirstMissing(list: List<Int>, start: Int, end: Int, network: String?): Int {
        val minNodeNumber: Int = getNodeNumberBoundaries(network).first
        val mid = (start + end) / 2
        // Left half has all elements from 0 to mid
        return when {
            list.isEmpty() -> minNodeNumber
            start > end -> end + 1 + minNodeNumber
            start + minNodeNumber != list[start] -> start + minNodeNumber
            list[mid] == mid + minNodeNumber -> findFirstMissing(list, mid + 1, end, network)

            // Left half has all elements from 0 to mid
            else -> findFirstMissing(list, start, mid, network)
        }

    }

    @Throws(NetworkNotFoundException::class)
    private fun getNodeNumberBoundaries(network: String?): Pair<Int, Int> {
        return if (networksConfiguration.map.containsKey(network)) {
            Pair(
                networksConfiguration.map[network]!!.minNodeNumber,
                networksConfiguration.map[network]!!.maxNodeNumber
            )
        } else {
            throw NetworkNotFoundException()
        }
    }
}